x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -129,83 +161,103 @@ func (k Keeper) UpdateDelegationBalances(ctx sdk.Context, hostZone types.HostZon
-			latestCompletionTime = undelegateResponse.CompletionTime
+		responseCompletionTime := uint64(undelegateResponse.CompletionTime.UnixNano())
+		if responseCompletionTime > latestCompletionTime {
+			latestCompletionTime = responseCompletionTime
 		}
 	}
 
-	if latestCompletionTime.IsZero() {
-		return nil, errorsmod.Wrapf(types.ErrInvalidPacketCompletionTime, "Invalid completion time (%s) from txMsg", latestCompletionTime.String())
+	if latestCompletionTime == 0 {
+		return 0, errorsmod.Wrapf(types.ErrInvalidPacketCompletionTime, "Invalid completion time 0 from txMsg")
 	}
-	return &latestCompletionTime, nil
+	return latestCompletionTime, nil
 }
 
-// UpdateHostZoneUnbondings does two things:
-//  1. Update the time of each hostZoneUnbonding on each epochUnbondingRecord
-//  2. Return the number of stTokens that need to be burned
-func (k Keeper) UpdateHostZoneUnbondings(
+// Updates the host zone unbonding records after a successful undelegation batch
+// The StTokensToBurn and the NativeTokensToUnbond amounts on the records are
+// decremented in a cascading fashion starting from the earliest record
+// The latest completion times is also set on each record if the time from the
+// batch is later than what's currently on the record
+func (k Keeper) UpdateHostZoneUnbondingsAfterUndelegation(
 	ctx sdk.Context,
-	latestCompletionTime time.Time,
 	chainId string,
-	undelegateCallback types.UndelegateCallback,
-) (stTokenBurnAmount sdkmath.Int, err error) {
-	stTokenBurnAmount = sdkmath.ZeroInt()
-	for _, epochNumber := range undelegateCallback.EpochUnbondingRecordIds {
-		epochUnbondingRecord, found := k.RecordsKeeper.GetEpochUnbondingRecord(ctx, epochNumber)
+	epochUnbondingRecordIds []uint64,
+	totalStTokensToBurn sdkmath.Int,
+	totalNativeTokensUnbonded sdkmath.Int,
+	unbondingTime uint64,
+) error {
+	// Loop each epoch unbonding record starting from the earliest
+	for _, epochNumber := range epochUnbondingRecordIds {
+		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochNumber, chainId)
 		if !found {
-			errMsg := fmt.Sprintf("Unable to find epoch unbonding record for epoch: %d", epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+			return errorsmod.Wrapf(recordstypes.ErrHostUnbondingRecordNotFound,
+				"host zone unbonding not found for epoch %d and %s", epochNumber, chainId)
 		}
-		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochUnbondingRecord.EpochNumber, chainId)
-		if !found {
-			errMsg := fmt.Sprintf("Host zone unbonding not found (%s) in epoch unbonding record: %d", chainId, epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+
+		// Determine the amount to decrement from each record, flooring at 0
+		stTokensToBurn := sdkmath.MinInt(hostZoneUnbonding.StTokensToBurn, totalStTokensToBurn)
```

Comment:
Hold on, explain this logic for me in more detail please.
What if we have this situation and the middle batch2 is calledback first:  
- batch1 epochIds [1, 1, 1, 1, 1]
- batch2 epochids [1, 1, 2, 2, 2]
- batch3 epochids [2, 2, 2, 2, 2]
Then when processing epochid1 the totalStTokensToBurn would be the sum across the [1, 1, 2, 2, 2] and the hostZoneUnbonding.stTokensToBurn would be the sum across the [1, 1, 1, 1, 1] [1, 1] but for this you really only want the sum across [1, 1] the sum across [1, 1, 1, 1, 1] will be handled in a different callback...

Reply:
I think you might need to pass in the splitundelegations which were part of the batch to compute this accurately not just the total sums but I could be missing something.

Reply:
so instead, you would just have [1,2] in the callback data for each batch. The batch will always reference all EUR's that were included in the total sum of native tokens that needed to be unbonded

I'm not sure if that perfectly answer your question, but if you're still confused after we can hop on a quick call

Reply:
I have the same question as Ethan.

>  The batch will always reference all EUR's that were included in the total sum of native tokens that needed to be unbonded

It seems like you're saying the CB from batch2 would process all 1s and all 2s but my understanding is that in this case we would only want to handle [1,1] and [2,2,2].

Are there unit tests for a case like this that I could read to help me understand?

Reply:
Thanks that helps. Consider this resolved from my end.